### ISSUE TITLE
ECONNRESET

### ISSUE THREAD

**Issue by Remixful** [discordjs/discord.js#2227]
**Please describe the problem you are having in as much detail as possible:**
```
Error: read ECONNRESET
at exports._errnoException (util.js:1022:11)
at TLSWrap.onread (net.js:569:26)
```
I've tried catching it in client disconnect and error events, yet they both don't seem to be getting called.
Supposedly this was supposed to be fixed over a year ago in: https://github.com/hydrabolt/discord.js/issues/694

**Further details:**

- discord.js version: 11.2.1
- node.js version: v8.9.4
- Operating system: Windows 10
- Priority this issue should have – please be realistic and elaborate if possible:
This causes my bot to crash constantly and I cannot restart it, even with pm2. So pretty high for me.

**Comment by kyranet** [comment]
I can reproduce this issue sometimes, I have got around ~20 during the last 30 days, however:

> "ECONNRESET" means the other side of the TCP conversation abruptly closed its end of the connection. This is most probably due to one or more application protocol errors. You could look at the API server logs to see if it complains about something.

> What could also be the case: at random times, the other side is overloaded and simply kills the connection as a result. If that's the case, depends on what you're connecting to exactly…

> But one thing's for sure: you indeed have a read error on your TCP connection which causes the exception. You can see that by looking at the error code you posted in your edit, which confirms it.

Read more in [StackOverflow](https://stackoverflow.com/questions/17245881/node-js-econnreset).

**Comment by hydrabolt** [comment]
This is hopefully probably fixed in the voice rewrite branch, if you could both give it a go that'd be cool! Although it branches from master, so if you're not using master you might have to refactor some code.

**Comment by kyranet** [comment]
@hydrabolt it's not fixed. But it is true the voice-rewrite branch works much better. In the same host, I'm running two bots, Skyra (using latest master) and Sneyra (using latest voice-rewrite), in a frame of 30 days, Skyra got around 20 `ECONNRESET` errors while Sneyra... only two (but with `EPIPE` error aswell).

In "harder" cases (when the API's ping is too high, when Discord is performing quite bad) Skyra tends to silently shutdown, appearing online but unresponsive (probably the websocket disconnected?), but Sneyra stays online and reconnects correctly.

![image](https://user-images.githubusercontent.com/24852502/34570189-555bca02-f16b-11e7-800f-53f2c2fbf1b1.png)

**To sum up**: The voice-rewrite fixes and makes these errors less frequent, but does not fix them totally.

**Comment by hydrabolt** [comment]
The issues happen and they're problems with the streams you're passing in, not discord.js (as far as I can tell), so as long as discord.js doesn't crash because of these errors I don't think it's an issue.

**Comment by appellation** [comment]
this doesn't seem to be an issue with voice necessarily, since voice only uses TCP for the websocket connection. without proper stack traces it's hard to debug the source of errors such as these; however, if they are indeed from a discord.js TCP connection, it would be useful if the library could handle them gracefully. also @hydrabolt, the issue author mentioned that his client is crashing: but, while I can confirm that this does issue is happening, I can't confirm whether this causes my client to reboot as well.

**Comment by hydrabolt** [comment]
It should definitely be fixed in voice-rewrite, not too sure about 11.2.1.

**Comment by y3nd** [comment]
I'm on 11.3 and I'm getting `Error: read ECONNRESET` each ~30 hours, matching with @kyranet 
I'll try to use master branch next week.

**Comment by Jisagi** [comment]
I also get this on latest master. It's rare but it happens.
I don't use any kind of voice features.

**Comment by AGuyNamedJens** [comment]
Yeah i was googling for a fix for this ECONRESET, so far nothing found. and same as @y3nd. On my raspberry pi my bot (running master) gets this ECONRESET after ~30 hours too. It's really frustrating since my pi literally crashes connection to the internet when the error happends..

**Comment by Lewdcario** [comment]
In response to @AGuyNamedJens, you can simply attach an `error` event listener in order to prevent crashes in the meantime (and get a nicer output possibly)

**Comment by AGuyNamedJens** [comment]
@Lewdcario i already have an error event, but the problem is the ECONRESET isnt catched and cuts off my entire pi's conbection to the internet

**Comment by Yaume230** [comment]
I get it at random moment, 10 hours or 3 days, what i done : 
reinstall my raspberry pi, seems no more ECONNRESET

**Comment by AGuyNamedJens** [comment]
i have too much on my pi to reset it, and this econreset happends since i bought the pi starterkit..

**Comment by hydrabolt** [comment]
Can you still reproduce this issue? Also, are you catching `error` from StreamDispatchers and VoiceConnections?

**Comment by hydrabolt** [comment]
This should be fixed now (as in the error may occur, but the process shouldn't die), if it's not please reopen an issue.

**Comment by hasezoey** [comment]
i get randomly mostly the same Error (at "2018-04-12T15:59:34.192Z")
```
{ Error: Unhandled "error" event. ([object Object])
    at Client.emit (events.js:186:19)
    at WebSocketConnection.onError (R:\Visual Studio\Projects\sylvi_remake\node_modules\discord.js\src\client\websocket\WebSocketConnection.js:374:17)
    at WebSocket.onError (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\event-target.js:128:16)
    at emitOne (events.js:116:13)
    at WebSocket.emit (events.js:211:7)
    at _receiver.cleanup (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\websocket.js:211:14)
    at Receiver.cleanup (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\receiver.js:557:13)
    at WebSocket.finalize (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\websocket.js:206:20)
    at emitOne (events.js:116:13)
    at TLSSocket.emit (events.js:211:7)
  context:
   ErrorEvent {
     target:
      WebSocket {
        domain: null,
        _events: [Object],
        _eventsCount: 4,
        _maxListeners: undefined,
        readyState: 2,
        protocol: '',
        _binaryType: 'nodebuffer',
        _finalize: [Function: bound finalize],
        _closeFrameReceived: false,
        _closeFrameSent: false,
        _closeMessage: '',
        _closeTimer: null,
        _finalized: true,
        _closeCode: 1006,
        _extensions: {},
        _isServer: false,
        _receiver: [Object],
        _sender: [Object],
        _socket: [Object],
        _error: null,
        url: 'wss://gateway.discord.gg/?v=6&encoding=etf',
        _req: null },
     type: 'error',
     message: 'read ECONNRESET',
     error: { Error: read ECONNRESET
    at _errnoException (util.js:1022:11)
    at TLSWrap.onread (net.js:628:25) code: 'ECONNRESET', errno: 'ECONNRESET', syscall: 'read' } } }
```
(using newset discord.js from npm)

**Comment by blockba5her** [comment]
+1 above I get the same error.

**Comment by SpaceEEC** [comment]
You both probably want to handle the client's error event then.

**Comment by hasezoey** [comment]
i handle the client errors (console.log) and still get:
```
{ Error: Unhandled "error" event. ([object Object])
    at VoiceWebSocket.emit (events.js:186:19)
    at VoiceWebSocket.onError (R:\Visual Studio\Projects\sylvi_remake\node_modules\discord.js\src\client\voice\VoiceWebSocket.js:156:10)
    at WebSocket.onError (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\event-target.js:128:16)
    at emitOne (events.js:116:13)
    at WebSocket.emit (events.js:211:7)
    at _receiver.cleanup (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\websocket.js:211:14)
    at Receiver.cleanup (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\receiver.js:535:15)
    at WebSocket.finalize (R:\Visual Studio\Projects\sylvi_remake\node_modules\ws\lib\websocket.js:206:20)
    at emitOne (events.js:116:13)
    at TLSSocket.emit (events.js:211:7)
  context:
   ErrorEvent {
     target:
      WebSocket {
        domain: null,
        _events: [Object],
        _eventsCount: 4,
        _maxListeners: undefined,
        readyState: 2,
        protocol: '',
        _binaryType: 'nodebuffer',
        _finalize: [Function: bound finalize],
        _closeFrameReceived: true,
        _closeFrameSent: true,
        _closeMessage: '',
        _closeTimer: [Object],
        _finalized: true,
        _closeCode: 1005,
        _extensions: {},
        _isServer: false,
        _receiver: [Object],
        _sender: [Object],
        _socket: [Object],
        _error: null,
        url: 'wss://eu-central124.discord.gg',
        _req: null },
     type: 'error',
     message: 'read ECONNRESET',
     error: { Error: read ECONNRESET
    at _errnoException (util.js:1022:11)
    at TLSWrap.onread (net.js:628:25) code: 'ECONNRESET', errno: 'ECONNRESET', syscall: 'read' } } }
````

**Comment by hydrabolt** [comment]
can you also add error handlers for your voice connections and any dispatchers you have?

**Comment by atomheartother** [comment]
Is this supposed to be fixed? I get the same thing, my server/bot seems to die on ECONNRESET. Yes, I have a handler for the exception, it logs it, and then the process still crashes... Unless I'm missing something?

Error handling code:
```
dClient.on('error', (error) => {
    console.error(new Date() + ": Discord client encountered an error");
    console.error(error);
});```

Log:
```
Fri Sep 07 2018 13:39:39 GMT+0200 (CEST): Discord client encountered an error
ErrorEvent {
  target: 
   WebSocket {
     _events: 
      { message: [Function],
        open: [Function],
        error: [Function],
        close: [Function] },
     _eventsCount: 4,
     _maxListeners: undefined,
     readyState: 2,
     protocol: '',
     _binaryType: 'nodebuffer',
     _finalize: [Function: bound finalize],
     _closeFrameReceived: false,
     _closeFrameSent: false,
     _closeMessage: '',
     _closeTimer: null,
     _finalized: true,
     _closeCode: 1006,
     _extensions: {},
     _isServer: false,
     _receiver: 
      Receiver {
        _binaryType: 'nodebuffer',
        _extensions: null,
        _maxPayload: 0,
        _bufferedBytes: 0,
        _buffers: null,
        _compressed: false,
        _payloadLength: 36,
        _fragmented: 0,
        _masked: false,
        _fin: true,
        _mask: null,
        _opcode: 1,
        _totalPayloadLength: 0,
        _messageLength: 0,
        _fragments: null,
        _cleanupCallback: null,
        _isCleaningUp: false,
        _hadError: false,
        _loop: false,
        add: [Function: bound add],
        onmessage: null,
        onclose: null,
        onerror: null,
        onping: null,
        onpong: null,
        _state: 0 },
     _sender: 
      Sender {
        _extensions: {},
        _socket: [TLSSocket],
        _firstFragment: true,
        _compress: false,
        _bufferedBytes: 0,
        _deflating: false,
        _queue: [] },
     _socket: 
      TLSSocket {
        _tlsOptions: [Object],
        _secureEstablished: true,
        _securePending: false,
        _newSessionPending: false,
        _controlReleased: true,
        _SNICallback: null,
        servername: 'gateway.discord.gg',
        alpnProtocol: false,
        authorized: true,
        authorizationError: null,
        encrypted: true,
        _events: [Object],
        _eventsCount: 4,
        connecting: false,
        _hadError: false,
        _handle: null,
        _parent: null,
        _host: 'gateway.discord.gg',
        _readableState: [ReadableState],
        readable: false,
        _maxListeners: undefined,
        _writableState: [WritableState],
        writable: false,
        allowHalfOpen: false,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: undefined,
        _server: null,
        ssl: null,
        _requestCert: true,
        _rejectUnauthorized: true,
        parser: null,
        _httpMessage: null,
        [Symbol(res)]: [TLSWrap],
        [Symbol(asyncId)]: 10135705,
        [Symbol(lastWriteQueueSize)]: 0,
        [Symbol(timeout)]: null,
        [Symbol(kBytesRead)]: 795658,
        [Symbol(kBytesWritten)]: 25420,
        [Symbol(connect-options)]: [Object] },
     _error: null,
     url: 'wss://gateway.discord.gg/?v=6&encoding=json',
     _req: null },
  type: 'error',
  message: 'read ECONNRESET',
  error: { Error: read ECONNRESET
    at TLSWrap.onread (net.js:659:25) errno: 'ECONNRESET', code: 'ECONNRESET', syscall: 'read' } }```

I even have a small bit of code that's supposed to catch unhandled exceptions, nothing will do. The process seems to just crash after ECONNRESET, not only that but I also have to reboot the server after this... I have no voice connections nor any dispatchers.

**Comment by char2** [comment]
> Is this supposed to be fixed? I get the same thing, my server/bot seems to die on ECONNRESET. Yes, I have a handler for the exception, it logs it, and then the process still crashes... Unless I'm missing something?
> 
> Error handling code:
> 
> ```
> dClient.on('error', (error) => {
>     console.error(new Date() + ": Discord client encountered an error");
>     console.error(error);
> });
> ```
> Log:
> 
> ```
> Fri Sep 07 2018 13:39:39 GMT+0200 (CEST): Discord client encountered an error
> ErrorEvent {
>   target: 
>    WebSocket {
>      _events: 
>       { message: [Function],
>         open: [Function],
>         error: [Function],
>         close: [Function] },
>      _eventsCount: 4,
>      _maxListeners: undefined,
>      readyState: 2,
>      protocol: '',
>      _binaryType: 'nodebuffer',
>      _finalize: [Function: bound finalize],
>      _closeFrameReceived: false,
>      _closeFrameSent: false,
>      _closeMessage: '',
>      _closeTimer: null,
>      _finalized: true,
>      _closeCode: 1006,
>      _extensions: {},
>      _isServer: false,
>      _receiver: 
>       Receiver {
>         _binaryType: 'nodebuffer',
>         _extensions: null,
>         _maxPayload: 0,
>         _bufferedBytes: 0,
>         _buffers: null,
>         _compressed: false,
>         _payloadLength: 36,
>         _fragmented: 0,
>         _masked: false,
>         _fin: true,
>         _mask: null,
>         _opcode: 1,
>         _totalPayloadLength: 0,
>         _messageLength: 0,
>         _fragments: null,
>         _cleanupCallback: null,
>         _isCleaningUp: false,
>         _hadError: false,
>         _loop: false,
>         add: [Function: bound add],
>         onmessage: null,
>         onclose: null,
>         onerror: null,
>         onping: null,
>         onpong: null,
>         _state: 0 },
>      _sender: 
>       Sender {
>         _extensions: {},
>         _socket: [TLSSocket],
>         _firstFragment: true,
>         _compress: false,
>         _bufferedBytes: 0,
>         _deflating: false,
>         _queue: [] },
>      _socket: 
>       TLSSocket {
>         _tlsOptions: [Object],
>         _secureEstablished: true,
>         _securePending: false,
>         _newSessionPending: false,
>         _controlReleased: true,
>         _SNICallback: null,
>         servername: 'gateway.discord.gg',
>         alpnProtocol: false,
>         authorized: true,
>         authorizationError: null,
>         encrypted: true,
>         _events: [Object],
>         _eventsCount: 4,
>         connecting: false,
>         _hadError: false,
>         _handle: null,
>         _parent: null,
>         _host: 'gateway.discord.gg',
>         _readableState: [ReadableState],
>         readable: false,
>         _maxListeners: undefined,
>         _writableState: [WritableState],
>         writable: false,
>         allowHalfOpen: false,
>         _sockname: null,
>         _pendingData: null,
>         _pendingEncoding: '',
>         server: undefined,
>         _server: null,
>         ssl: null,
>         _requestCert: true,
>         _rejectUnauthorized: true,
>         parser: null,
>         _httpMessage: null,
>         [Symbol(res)]: [TLSWrap],
>         [Symbol(asyncId)]: 10135705,
>         [Symbol(lastWriteQueueSize)]: 0,
>         [Symbol(timeout)]: null,
>         [Symbol(kBytesRead)]: 795658,
>         [Symbol(kBytesWritten)]: 25420,
>         [Symbol(connect-options)]: [Object] },
>      _error: null,
>      url: 'wss://gateway.discord.gg/?v=6&encoding=json',
>      _req: null },
>   type: 'error',
>   message: 'read ECONNRESET',
>   error: { Error: read ECONNRESET
>     at TLSWrap.onread (net.js:659:25) errno: 'ECONNRESET', code: 'ECONNRESET', syscall: 'read' } }
> ```
> I even have a small bit of code that's supposed to catch unhandled exceptions, nothing will do. The process seems to just crash after ECONNRESET, not only that but I also have to reboot the server after this... I have no voice connections nor any dispatchers.

I get the same issue exactly currently. Was there ever a fix for this?

**Comment by AGuyNamedJens** [comment]
Well the only fix i know for this is, Your Internet

The ECONRESET means the process lost connection to the internet for too long, causing it to die

The fix for me was making a direct ethernet cable from my router to my raspberry pi

**Comment by char2** [comment]
This would definitely explain why it sometimes happens much more frequently than other times.
I'm using an amazon EC2 server so the internet connection shouldn't be an issue.. perhaps it has intermittent network problems..

**Comment by atomheartother** [comment]
I'm also using a remote server, it's hosted at OVH... I am having to restart the server every few weeks pretty consistently because of this bug.I just have trouble believing that the internet connection is problem here for a provider as reputable as ovh, same for aws...

**Comment by char2** [comment]
Yes the same happens to me.. except probably once or twice a day.
Sometimes it happens many times in a row every few minutes.
Hopefully we get a way to troubleshoot this further.

**Comment by Sickness-chan** [comment]
Same problem for me, I host at OVH and I'm sure there's no network problems. But I keep getting this ECONNRESET error, trying to catch it, but it crashes the application anyway. I run in PM2, so the crashing doesn't really matter as it automatically restarts, but still.

**Comment by neilzar** [comment]
I'm experiencing this error every day, running on Ubuntu 18.04.1 LTS. These are the final three lines of the error:
```
error: { Error: read ECONNRESET
  at _errnoException (util.js:1022:11)
  at TLSWrap.onread (net.js:628:25) code: 'ECONNRESET', errno: 'ECONNRESET', syscall: 'read' } }
```
The rest of the error that came before it is (almost) identical to atomheartother's error.

I'm running `discord.js@11.4.2`, and I am not using any voice or dispatcher stuff. Just replying to chat messages and doing some MySQL stuff.

**Comment by Sv443** [comment]
It's still not fixed, I'm using the latest discord.js version (`11.4.2`) and still get about one `read ECONNRESET` error each day. I just made it ignore that error but that's just a workaround and I'd want to avoid the error altogether.

**Comment by Kos-M** [comment]
@Sv443  Had same expirience.
bot logout on conreset error. after that is loged all the time.
i solved  adding : 
```javascript
 process.on('uncaughtException', function (err) {
  console.log(err);
})

**Comment by Sv443** [comment]
> 
> 
> @Sv443 Had same expirience.
> bot logout on conreset error. after that is loged all the time.
> i solved adding :
> 
> ```js
>  process.on('uncaughtException', function (err) {
>   console.log(err);
> })
> ```

But I don't want to have a workaround, I want the bug to be fixed altogether.

**Comment by Kos-M** [comment]
@Sv443   i see..  that exception is trigered when the other end ( some discord server i suppose)  resets the connection.
Without the handle `uncaughtException` app crashes.

**Comment by Sv443** [comment]
@Kos-M yeah. It helps that I made [this package](https://npmjs.com/package/node-wrap) to restart the bot once it crashes but the error handler makes it even better.

**Comment by Kos-M** [comment]
Nice touch too, best solution is to fix that in discord.js , im not a contributor to this project ,only using it for a project.
And i found that way to protect app from crash.
I Hope will fixed natively..

**Comment by SoundTest** [comment]
this suddenly started happening to me, no clue why

**Comment by Sv443** [comment]
> 
> 
> this suddenly started happening to me, no clue why

The solution is to add an error handler. There's still no news from the Discord API.
